keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,392 @@
+to restart the entire JobSet and resume training from the latest checkpoint.
+
+**Example Failure Policy configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: configurable-failure-policy-2
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: {{topologyDomain}} # 1:1 job replica to topology domain assignment
+spec:
+  # Failure Policy configured to restart any failed child job in the cpu-worker ReplicatedJob, but fall back to the
+  # default behavior of restarting the entire JobSet if any child job in the tpu-worker ReplicatedJob fails.
+  failurePolicy:
+    rules:
+    - action: RestartJob
+      targetReplicatedJobs: cpu-worker
+    maxRestarts: 10
+  replicatedJobs:
+  - name: gpu-worker 
+    replicas: 4 # set to number of node pools
+    template:
+      spec:
+        parallelism: 2 # number of nodes per pool
+        completions: 2 # number of nodes per pool
+        backoffLimit: 0
+        template:
+          spec:
+            containers:
+            - name: main
+              image: pytorch:latest
+              command: ["..."]
+            resources:
+              limits:
+                nvidia.com/gpu: 1
+  - name: gpu-worker 
+    replicas: 1 # set to number of node pools
+    template:
+      spec:
+        parallelism: 1 # number of nodes per pool
+        completions: 1 # number of nodes per pool
+        backoffLimit: 0
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+              
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Failure Policy API
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+  // Fail JobSet immediately, regardless of maxRestarts.
+  FailJobSet FailurePolicyAction = "FailJobSet"
+
+  // Don't count the failure against maxRestarts.
+  Ignore FailurePolicyAction = "Ignore"
+
+  // The failed child job is permanently failed and should not be
+  // restarted, active child jobs will continue to run to completion.
+  FailJob FailurePolicyAction = "FailJob"
```

Comment:
ditto, for each action, we should have a story, if we don't have a clear one, we can hold off adding it.

Reply:
Ok makes sense, I've added 3 user stories, one for each of the following actions:

- `FailJobSet`
- `Ignore`
- `RestartReplicatedJob`

However, I haven't been able to think of a concrete use case for `FailJob` or `RestartJob`. It seems ReplicatedJob level granularity is sufficient for the use cases I have thought of so far. 

I asked @vsoch about potential HPC use cases, to better understand the requirements there, but based on our initial conversation I think the 3 actions above are sufficient.

If anyone has any concrete use cases for restarting an individual child job (or leaving it failed), please feel free to add a coment.

Reply:
`FailJob` or `RestartJob` might be relevant if you are running some kind of simulation (a number of replicas within a ReplicatedJob) that would generate a result that is valid or invalid, and you had a way to measure the distance from that. For example, if each Job within the Replicated Job generates a random parameter matrix for the simulation within that Job (to produce a result that is either valid or invalid) if your resulting value is too far off from some metric of goodness you would want to fail that specific Job to generate a new parameter matrix - the one you have isn't good. If the result is within a reasonable distance of your metric of goodness (close to valid, but still invalid) you might want to `RestartJob` because you have some confidence it could produce a valid result. You'd want to run the simulation again.

Reply:
> `FailJob` or `RestartJob` might be relevant if you are running some kind of simulation (a number of replicas within a ReplicatedJob) that would generate a result that is valid or invalid, and you had a way to measure the distance from that. For example, if each Job within the Replicated Job generates a random parameter matrix for the simulation within that Job (to produce a result that is either valid or invalid) if your resulting value is too far off from some metric of goodness you would want to fail that specific Job to generate a new parameter matrix - the one you have isn't good. If the result is within a reasonable distance of your metric of goodness (close to valid, but still invalid) you might want to `RestartJob` because you have some confidence it could produce a valid result. You'd want to run the simulation again.

How would this metric be surfaced to the JobSet controller though, in order to determine if it should FailJob or RestartJob? It only can look at failed Jobs and examine the reason it failed.

Reply:
It wouldn't - it would be internal to the application, which would then exit with the appropriate code to determine the desired action.

Reply:
> It wouldn't - it would be internal to the application, which would then exit with the appropriate code to determine the desired action.

Right, but the podFailurePolicy would be configured to do what based on the container exit code?

The only podFailurePolicy options are to Ignore or FailJob right? If I understand correctly, FailJob is what the podFailurePolicy would be applying here, regardless of whether the application exit code indicates the job should restart or stay failed. 

The "reason" on the failure condition applied to the Job will be `PodFailurePolicy` regardless of which container exit code caused the podFailurePolicy to fail the job, so the JobSet controller won't know whether to RestartJob or FailJob. Am I missing something?

Reply:
The replicatedJob has multiple jobs under it, each generated from a template. If the top level replicatedJob has a global Failure or Restart policy, that means it doesn't vary on the level of the container. But what we need:
 - if JobA under the replicated Job (let's say it's one pod) has exit codes A-C, I want FailJob. I don't want that particular Job (pod) to restart, I know it's a doozie.
 - If JobB under the replicated Job  (also one pod) has exit codes D-F, that's a signal from the application to try again. I restart the Job. 
 
The above is mapping 1 container --> 1 pod --> 1 node, which is what we often do for our workloads because we need the full resources, in case that is relevant. So when I talk about 1 Job I am talking about one Pod and one container.

If I'm still not understanding then I'm having a hard time understanding the many levels of this proposal, and that's probably a bad thing from a user perspective.

Reply:
Another concern I have (that I've expressed before) is where to draw the line between what a workflow manager is supposed to do (that has a DAG, monitors statuses and makes decisions) vs. what Kubernetes is doing. We keep stepping into representing functionality that (traditionally) a workflow manager would handle and I'm not sure where the line is being drawn. In situations like these there should be enough information provided about any kind of status for the workload manager to make the right decision, and it's up to the abstraction to just provide that (but not to act like a workflow manager).

Reply:
Yeah, I understand your use case, but the just the limitations of the Job controller and PodFailurePolicies don't allow us to support this in JobSet yet. JobA and JobB would need distinct JobFailureReasons in order for JobSet controller to apply different actions to them. However, the Job controller currently only supports a single Job failure reason for podFaliurePolicies ([link](https://sourcegraph.com/github.com/kubernetes/kubernetes@f7f1ea6fa23ec0fc527ff10174bbe0dc3bec77da/-/blob/staging/src/k8s.io/api/batch/v1/types.go?L542:37-542:55)), which are the only mechanism the Job controller has for distinguishing between different application exit codes. So the underlying infrastructure in the Job API just doesn't exist yet to support this, unfortunately.

Reply:
okay understood - so the use case is valid, but not possible yet. And it's not that you don't see a case for `FailJob` or `RestartJob` but rather that there is no way to actually implement it. I just want to clarify that, because those are two different things (although they result in the same outcome that we can't add the cases here yet).

Reply:
The most important thing is whether the API we are proposing now allows supporting this case in the future when the enhancement to the Job API is implemented: https://github.com/kubernetes/kubernetes/issues/122972

If we agree, then I think we can resolve this, right?

Reply:
Ok that makes sense, we can add FailJob and RestartJob proactively in anticipation of the Job API enhancements. Are we fairly sure https://github.com/kubernetes/kubernetes/issues/122972 will be approved and implemented? If so, I will work on adding some user stories for FailJob and RestartJob, based on the use case @vsoch described.

Reply:
I don't think we can be sure of anything, but if it's a valid and needed use case, we should continue advocating for it (I certainly will)!

Reply:
Even if 1229721 will be implemented (we would probably need a KEP in the next week) for 1.30. We still would have to support previous versions without this change so I think it would be best to implement without it and then maybe create a follow up issue once 1229721 is implemented and all supported releases of kube have this change.

Reply:
> If so, I will work on adding some user stories for FailJob and RestartJob, based on the use case @vsoch described.

We can do that citing the open issue as a dependency.

Reply:
Ok I added user story 4 which covers the FailJob and RestartJob use cases, and cited https://github.com/kubernetes/kubernetes/issues/122972 as a dependency.